### PR TITLE
Implement SSO In Adaptive Cards

### PR DESCRIPTION
**Pull request by Harikrishnan-MSFT**

### CODE REVIEW

**Comment by ChetanSharma-msft** [comment]
Executed build pipeline:

File: samples/bot-conversation-sso-quickstart/csharp_dotnetcore/BotConversationSsoQuickstart/Bots/DialogBot.cs [hunk 1]
```diff
@@ -23,6 +32,7 @@ public class DialogBot<T> : TeamsActivityHandler where T : Dialog
         protected readonly Dialog _dialog;
         protected readonly ILogger _logger;
         protected readonly BotState _userState;
+        private readonly string _connectionName = "UniversalAdaptiveCardsConnection"; //only testing perforce
```
Comment (ChetanSharma-msft) [inline samples/bot-conversation-sso-quickstart/csharp_dotnetcore/BotConversationSsoQuickstart/Bots/DialogBot.cs:35]: Why we need this comment ?

File: samples/bot-conversation-sso-quickstart/csharp_dotnetcore/BotConversationSsoQuickstart/Bots/DialogBot.cs [hunk 2]
```diff
@@ -32,18 +42,14 @@ public DialogBot(ConversationState conversationState, UserState userState, T dia
...
-        /// <param name="cancellationToken">Propagates notification that operations should be canceled.</param>
-        /// <returns>A task that represents the work queued to execute.</returns>
-        /// <remarks>
-        /// Reference link: https://docs.microsoft.com/en-us/dotnet/api/microsoft.bot.builder.activityhandler.onturnasync?view=botbuilder-dotnet-stable.
-        /// </remarks>
-        public override async Task OnTurnAsync(
-            ITurnContext turnContext,
-            CancellationToken cancellationToken = default(CancellationToken))
+        private async Task<string> GetSignInLinkAsync(ITurnContext turnContext, CancellationToken cancellationToken)
+        {
```
Comment (ChetanSharma-msft) [inline samples/bot-conversation-sso-quickstart/csharp_dotnetcore/BotConversationSsoQuickstart/Bots/DialogBot.cs:46]: Please add summary for this method.

File: samples/bot-conversation-sso-quickstart/csharp_dotnetcore/BotConversationSsoQuickstart/Bots/DialogBot.cs [hunk 3]
```diff
@@ -52,23 +58,160 @@ public DialogBot(ConversationState conversationState, UserState userState, T dia
...
-        /// <param name="cancellationToken">Propagates notification that operations should be canceled.</param>
-        /// <returns>A Task resolving to either a login card or the adaptive card of the Reddit post.</returns>
-        /// <remarks>
-        /// For more information on bot messaging in Teams, see the documentation
-        /// https://docs.microsoft.com/en-us/microsoftteams/platform/bots/how-to/conversations/conversation-basics?tabs=dotnet#receive-a-message .
-        /// </remarks>
-        protected override async Task OnMessageActivityAsync(
-            ITurnContext<IMessageActivity> turnContext,
-            CancellationToken cancellationToken)
+        protected override async Task OnMessageActivityAsync(ITurnContext<IMessageActivity> turnContext, CancellationToken cancellationToken)
```
Comment (ChetanSharma-msft) [inline samples/bot-conversation-sso-quickstart/csharp_dotnetcore/BotConversationSsoQuickstart/Bots/DialogBot.cs:61]: Please add summary for this method.

File: samples/bot-conversation-sso-quickstart/csharp_dotnetcore/BotConversationSsoQuickstart/Bots/DialogBot.cs [hunk 4]
```diff
@@ -52,23 +58,160 @@ public DialogBot(ConversationState conversationState, UserState userState, T dia
...
+                var varInitialAdaptiveCard = GetFirstOptionsAdaptiveCard(striPath, signInLink, turnContext.Activity.From.Name, varMember.Id);
+                await turnContext.SendActivityAsync(MessageFactory.Attachment(varInitialAdaptiveCard), cancellationToken);
+            }
+            else
+            {
+                await turnContext.SendActivityAsync(MessageFactory.Text("Please send 'login' for options"), cancellationToken);
+            }
+        }
+
+        protected override async Task<InvokeResponse> OnInvokeActivityAsync(ITurnContext<IInvokeActivity> turnContext, CancellationToken cancellationToken)
```
Comment (ChetanSharma-msft) [inline samples/bot-conversation-sso-quickstart/csharp_dotnetcore/BotConversationSsoQuickstart/Bots/DialogBot.cs:84]: Please add summary for this method.

File: samples/bot-conversation-sso-quickstart/csharp_dotnetcore/BotConversationSsoQuickstart/Bots/DialogBot.cs [hunk 5]
```diff
@@ -52,23 +58,160 @@ public DialogBot(ConversationState conversationState, UserState userState, T dia
...
+                string verb = actiondata["verb"].ToString();
+                JObject authentication = null;
+
+                string state = null;
+                if (value["state"] != null)
+                {
+                    state = value["state"].ToString();
+                }
+                // authToken and state are absent, handle verb
+                if (authentication == null && state == null)
```
Comment (ChetanSharma-msft) [inline samples/bot-conversation-sso-quickstart/csharp_dotnetcore/BotConversationSsoQuickstart/Bots/DialogBot.cs:111]: One line space before it.

File: samples/bot-conversation-sso-quickstart/csharp_dotnetcore/BotConversationSsoQuickstart/Bots/DialogBot.cs [hunk 6]
```diff
@@ -52,23 +58,160 @@ public DialogBot(ConversationState conversationState, UserState userState, T dia
...
+                {
+                    switch (verb)
+                    {
+                        case "initiateSSO":
+                            return await initiateSSOAsync(turnContext, cancellationToken);
+                    }
+                }
+                // authToken or state is present. Verify token/state in invoke payload and return AC response
```
Comment (ChetanSharma-msft) [inline samples/bot-conversation-sso-quickstart/csharp_dotnetcore/BotConversationSsoQuickstart/Bots/DialogBot.cs:119]: We can move this comment inside else block

File: samples/bot-conversation-sso-quickstart/csharp_dotnetcore/BotConversationSsoQuickstart/Bots/DialogBot.cs [hunk 7]
```diff
@@ -52,23 +58,160 @@ public DialogBot(ConversationState conversationState, UserState userState, T dia
...
+                {
+                    return createAdaptiveCardInvokeResponseAsync(authentication, state);
+                }
+            }
+
+            return null;
+        }
+
+        private async Task<InvokeResponse> initiateSSOAsync(ITurnContext<IInvokeActivity> turnContext, CancellationToken cancellationToken)
+        {
```
Comment (ChetanSharma-msft) [inline samples/bot-conversation-sso-quickstart/csharp_dotnetcore/BotConversationSsoQuickstart/Bots/DialogBot.cs:130]: Add summary for this method.

File: samples/bot-conversation-sso-quickstart/csharp_dotnetcore/BotConversationSsoQuickstart/Bots/DialogBot.cs [hunk 8]
```diff
@@ -52,23 +58,160 @@ public DialogBot(ConversationState conversationState, UserState userState, T dia
...
+            var signInLink = await GetSignInLinkAsync(turnContext, cancellationToken).ConfigureAwait(false);
+            var oAuthCard = new OAuthCard
```
Comment (ChetanSharma-msft) [inline samples/bot-conversation-sso-quickstart/csharp_dotnetcore/BotConversationSsoQuickstart/Bots/DialogBot.cs:132]: One line space before it.

File: samples/bot-conversation-sso-quickstart/csharp_dotnetcore/BotConversationSsoQuickstart/Bots/DialogBot.cs [hunk 9]
```diff
@@ -52,23 +58,160 @@ public DialogBot(ConversationState conversationState, UserState userState, T dia
...
+                        {
+                            Type = ActionTypes.Signin,
+                            Value = signInLink,
+                            Title = "Please sign in",
+                        },
+                    }
+            };
+            var loginReqResponse = JObject.FromObject(new
+            {
+                statusCode = 401,
```
Comment (ChetanSharma-msft) [inline samples/bot-conversation-sso-quickstart/csharp_dotnetcore/BotConversationSsoQuickstart/Bots/DialogBot.cs:152]: One line space before it.

File: samples/bot-conversation-sso-quickstart/csharp_dotnetcore/BotConversationSsoQuickstart/Bots/DialogBot.cs [hunk 10]
```diff
@@ -52,23 +58,160 @@ public DialogBot(ConversationState conversationState, UserState userState, T dia
...
+                type = "application/vnd.microsoft.activity.loginRequest",
+                value = oAuthCard
+            });
+
+            return CreateInvokeResponse(loginReqResponse);
+        }
+
+        private InvokeResponse createAdaptiveCardInvokeResponseAsync(JObject authentication, string state, bool isBasicRefresh = false, string fileName = "adaptiveCardResponseJson.json")
```
Comment (ChetanSharma-msft) [inline samples/bot-conversation-sso-quickstart/csharp_dotnetcore/BotConversationSsoQuickstart/Bots/DialogBot.cs:160]: Add summary for this method.

File: samples/bot-conversation-sso-quickstart/csharp_dotnetcore/BotConversationSsoQuickstart/Bots/DialogBot.cs [hunk 11]
```diff
@@ -52,23 +58,160 @@ public DialogBot(ConversationState conversationState, UserState userState, T dia
...
+        {
+            //verify token is present or not
+
```
Comment (ChetanSharma-msft) [inline samples/bot-conversation-sso-quickstart/csharp_dotnetcore/BotConversationSsoQuickstart/Bots/DialogBot.cs:163]: // Verify if token is present or not.

File: samples/bot-conversation-sso-quickstart/csharp_dotnetcore/BotConversationSsoQuickstart/Bots/DialogBot.cs [hunk 12]
```diff
@@ -52,23 +58,160 @@ public DialogBot(ConversationState conversationState, UserState userState, T dia
...
+            bool isTokenPresent = authentication != null ? true : false;
+            bool isStatePresent = state != null && state != "" ? true : false;
+
+            // TODO : Use token or state to perform operation on behalf of user
```
Comment (ChetanSharma-msft) [inline samples/bot-conversation-sso-quickstart/csharp_dotnetcore/BotConversationSsoQuickstart/Bots/DialogBot.cs:167]: Why TODO?

File: samples/bot-conversation-sso-quickstart/csharp_dotnetcore/BotConversationSsoQuickstart/Bots/DialogBot.cs [hunk 13]
```diff
@@ -52,23 +58,160 @@ public DialogBot(ConversationState conversationState, UserState userState, T dia
...
+            AdaptiveCardTemplate template = new AdaptiveCardTemplate(adaptiveCardJson);
+            var authResultData = isTokenPresent ? "SSO success" : isStatePresent ? "OAuth success" : "SSO/OAuth failed";
+            if (isBasicRefresh)
+            {
+                authResultData = "Refresh done";
+            }
+            var payloadData = new
+            {
+                authResult = authResultData
+            };
```
Comment (ChetanSharma-msft) [inline samples/bot-conversation-sso-quickstart/csharp_dotnetcore/BotConversationSsoQuickstart/Bots/DialogBot.cs:181]: One line space before if and var

File: samples/bot-conversation-sso-quickstart/csharp_dotnetcore/BotConversationSsoQuickstart/Bots/DialogBot.cs [hunk 14]
```diff
@@ -52,23 +58,160 @@ public DialogBot(ConversationState conversationState, UserState userState, T dia
...
+
+            var adaptiveCardResponse = new AdaptiveCardInvokeResponse()
+            {
+                StatusCode = 200,
+                Type = AdaptiveCard.ContentType,
+                Value = JsonConvert.DeserializeObject(cardJsonstring)
+            };
+            return CreateInvokeResponse(adaptiveCardResponse);
+        }
+        private Attachment GetFirstOptionsAdaptiveCard(string[] filepath, string signInLink, string name = null, string userMRI = null)
```
Comment (ChetanSharma-msft) [inline samples/bot-conversation-sso-quickstart/csharp_dotnetcore/BotConversationSsoQuickstart/Bots/DialogBot.cs:193]: One line space before it and also add method summary

File: samples/bot-conversation-sso-quickstart/csharp_dotnetcore/BotConversationSsoQuickstart/Bots/DialogBot.cs [hunk 15]
```diff
@@ -52,23 +58,160 @@ public DialogBot(ConversationState conversationState, UserState userState, T dia
...
+            var card = JsonConvert.DeserializeObject<JObject>(cardJsonstring);
+            if (card["authentication"] != null && card["authentication"]["buttons"] != null && card["authentication"]["buttons"][0] != null)
+            {
+                card["authentication"]["buttons"][0]["value"] = signInLink;
+            }
+            var adaptiveCardAttachment = new Attachment()
+            {
+                ContentType = "application/vnd.microsoft.card.adaptive",
+                Content = card,
+            };
```
Comment (ChetanSharma-msft) [inline samples/bot-conversation-sso-quickstart/csharp_dotnetcore/BotConversationSsoQuickstart/Bots/DialogBot.cs:212]: Line spacing is not correct.

File: samples/bot-conversation-sso-quickstart/csharp_dotnetcore/BotConversationSsoQuickstart/Resources/ABSSORefresh.json [hunk 16]
```diff
@@ -0,0 +1,35 @@
...
+    {
+      "type": "TextBlock",
+      "text": "Welcome Universal Card",
+      "size": "Medium",
+      "weight": "Bolder"
+    }
+  ],
+  "$schema": "http://adaptivecards.io/schemas/adaptive-card.json",
+  "appId": "9b6dcf71-df47-4756-b486-47ca94dbacee",
+  "version": "1.4"
```
Comment (ChetanSharma-msft) [inline samples/bot-conversation-sso-quickstart/csharp_dotnetcore/BotConversationSsoQuickstart/Resources/ABSSORefresh.json:34]: Check if this version is correct and supported

File: samples/bot-conversation-sso-quickstart/csharp_dotnetcore/BotConversationSsoQuickstart/Resources/ABSSORefresh.json [hunk 17]
```diff
@@ -0,0 +1,35 @@
...
+    "text": "Signin Text",
+    "connectionName": "newConnection",
+    "tokenExchangeResource": {
+      "id": "9b6dcf71-df47-4756-b486-47ca94dbacee"
+    },
+    "buttons": [
+      {
+        "type": "signin",
+        "title": "Please sign in",
+        "value": "https://token.botframework.com/api/oauth/signin?signin=9b6dcf71-df47-4756-b486-47ca94dbacee"
```
Comment (ChetanSharma-msft) [inline samples/bot-conversation-sso-quickstart/csharp_dotnetcore/BotConversationSsoQuickstart/Resources/ABSSORefresh.json:20]: Why did we push app id?  9b6dcf71-df47-4756-b486-47ca94dbacee

Add place holder and update the readme file.

File: samples/bot-conversation-sso-quickstart/csharp_dotnetcore/BotConversationSsoQuickstart/Resources/ABSSORefresh.json [hunk 18]
```diff
@@ -0,0 +1,35 @@
...
+      "type": "Action.Execute",
+      "verb": "initiateSSO",
+      "title": "RefreshTitle"
+    }
+  },
+  "authentication": {
+    "text": "Signin Text",
+    "connectionName": "newConnection",
+    "tokenExchangeResource": {
+      "id": "9b6dcf71-df47-4756-b486-47ca94dbacee"
```
Comment (ChetanSharma-msft) [inline samples/bot-conversation-sso-quickstart/csharp_dotnetcore/BotConversationSsoQuickstart/Resources/ABSSORefresh.json:14]: Why did we push app id? 9b6dcf71-df47-4756-b486-47ca94dbacee

File: samples/bot-conversation-sso-quickstart/csharp_dotnetcore/BotConversationSsoQuickstart/Resources/adaptiveCardResponseJson.json [hunk 19]
```diff
@@ -0,0 +1,40 @@
...
+    {
+      "type": "TextBlock",
+      "text": "${authResult}",
+      "size": "Medium",
+      "weight": "Bolder"
+    }
+  ],
+  "$schema": "http://adaptivecards.io/schemas/adaptive-card.json",
+  "appId": "ecc9fe9e-03df-4597-9c7d-22828d713d68",
+  "version": "1.4"
```
Comment (ChetanSharma-msft) [inline samples/bot-conversation-sso-quickstart/csharp_dotnetcore/BotConversationSsoQuickstart/Resources/adaptiveCardResponseJson.json:39]: Check if this version is correct and supported

Also, why did we push app id ?

Add place holder and Update Readme file to

File: samples/bot-conversation-sso-quickstart/csharp_dotnetcore/BotConversationSsoQuickstart/Resources/options.json [hunk 20]
```diff
@@ -0,0 +1,31 @@
...
+              "value": "{\"bfKey\": \"bfVal\", \"conflictKey\": \"from value\"}"
+            }
+          }
+        }
+      ]
+    }
+  ],
+  "$schema": "http://adaptivecards.io/schemas/adaptive-card.json",
+  "appId": "9b6dcf71-df47-4756-b486-47ca94dbacee",
+  "version": "1.4"
```
Comment (ChetanSharma-msft) [inline samples/bot-conversation-sso-quickstart/csharp_dotnetcore/BotConversationSsoQuickstart/Resources/options.json:30]: Check if this version is correct and supported

Also, why did we push app id ?

Add place holder and Update Readme file to

**Review by ChetanSharma-msft — CHANGES_REQUESTED** [review]
Added my review comments.

**Comment by Pawan-MSFT** [comment]
Executed build pipeline:
https://github.com/OfficeDev/Microsoft-Teams-Samples/actions/runs/3821330043

Result:
image

File: samples/bot-conversation-sso-quickstart/csharp_dotnetcore/BotConversationSsoQuickstart/Bots/DialogBot.cs (hunk 1 above)
Comment (Harikrishnan-MSFT) [reply samples/bot-conversation-sso-quickstart/csharp_dotnetcore/BotConversationSsoQuickstart/Bots/DialogBot.cs]: Changes updated

File: samples/bot-conversation-sso-quickstart/csharp_dotnetcore/BotConversationSsoQuickstart/Bots/DialogBot.cs (hunk 2 above)
Comment (Harikrishnan-MSFT) [reply samples/bot-conversation-sso-quickstart/csharp_dotnetcore/BotConversationSsoQuickstart/Bots/DialogBot.cs:49]: Changes updated

File: samples/bot-conversation-sso-quickstart/csharp_dotnetcore/BotConversationSsoQuickstart/Bots/DialogBot.cs (hunk 3 above)
Comment (Harikrishnan-MSFT) [reply samples/bot-conversation-sso-quickstart/csharp_dotnetcore/BotConversationSsoQuickstart/Bots/DialogBot.cs:78]: Changes updated

File: samples/bot-conversation-sso-quickstart/csharp_dotnetcore/BotConversationSsoQuickstart/Bots/DialogBot.cs (hunk 4 above)
Comment (Harikrishnan-MSFT) [reply samples/bot-conversation-sso-quickstart/csharp_dotnetcore/BotConversationSsoQuickstart/Bots/DialogBot.cs:107]: Changes updated

File: samples/bot-conversation-sso-quickstart/csharp_dotnetcore/BotConversationSsoQuickstart/Bots/DialogBot.cs (hunk 5 above)
Comment (Harikrishnan-MSFT) [reply samples/bot-conversation-sso-quickstart/csharp_dotnetcore/BotConversationSsoQuickstart/Bots/DialogBot.cs:148]: Changes updated

File: samples/bot-conversation-sso-quickstart/csharp_dotnetcore/BotConversationSsoQuickstart/Resources/ABSSORefresh.json (hunk 16 above)
Comment (Harikrishnan-MSFT) [reply samples/bot-conversation-sso-quickstart/csharp_dotnetcore/BotConversationSsoQuickstart/Resources/ABSSORefresh.json]: Changes updated

File: samples/bot-conversation-sso-quickstart/csharp_dotnetcore/BotConversationSsoQuickstart/Resources/ABSSORefresh.json (hunk 17 above)
Comment (Harikrishnan-MSFT) [reply samples/bot-conversation-sso-quickstart/csharp_dotnetcore/BotConversationSsoQuickstart/Resources/ABSSORefresh.json]: Changes updated

File: samples/bot-conversation-sso-quickstart/csharp_dotnetcore/BotConversationSsoQuickstart/Resources/ABSSORefresh.json (hunk 18 above)
Comment (Harikrishnan-MSFT) [reply samples/bot-conversation-sso-quickstart/csharp_dotnetcore/BotConversationSsoQuickstart/Resources/ABSSORefresh.json]: Changes updated

File: samples/bot-conversation-sso-quickstart/csharp_dotnetcore/BotConversationSsoQuickstart/Resources/adaptiveCardResponseJson.json (hunk 19 above)
Comment (Harikrishnan-MSFT) [reply samples/bot-conversation-sso-quickstart/csharp_dotnetcore/BotConversationSsoQuickstart/Resources/adaptiveCardResponseJson.json]: Changes updated

File: samples/bot-conversation-sso-quickstart/csharp_dotnetcore/BotConversationSsoQuickstart/Resources/options.json (hunk 20 above)
Comment (Harikrishnan-MSFT) [reply samples/bot-conversation-sso-quickstart/csharp_dotnetcore/BotConversationSsoQuickstart/Resources/options.json]: Changes updated

**Comment by ChetanSharma-msft** [comment]
This is a duplicate PR for this: https://github.com/OfficeDev/Microsoft-Teams-Samples/pull/653

**Pull request closed without merging**